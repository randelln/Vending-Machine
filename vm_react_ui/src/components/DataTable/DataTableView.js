import React from 'react';
import { Table, Button } from 'reactstrap';
import { API_BASE_URL } from '../../constants';
import './DataTableStyles.css';

//Same as home container
export const checkout = (product, runningTotal, setLoadingBuy, setRunningTotal, fetchProducts, fetchCoins) => {
    if (product.price > runningTotal) {
        window.alert('Not enough funds');
    } else if (product.quantity == 0) {
        window.alert(`No ${product.name} avaliable. Coins have been refunded`);
    } else {
        setLoadingBuy(true);
        fetch(`${API_BASE_URL}/buy/${product.id}/${runningTotal}`, {
            method: 'get',
            headers: {
                'Content-Type': 'application/json'
            }
        })
            .then(res => res.json())
            .then(response => {
                setLoadingBuy(false);
                if (!response.error) {
                    window.alert(`Purchase of ${product.name} successful, coins refunded is R${response.balance}`);
                    setRunningTotal(0);
                    fetchProducts();
                    fetchCoins();
                } else {
                    window.alert('Purchase unsuccessful');
                    setRunningTotal(response.balance);
                }
            })
            .catch(err => {
                console.log(err);
            });
    }
};
// function that accepts these props
const DataTable = ({
                       products,
                       runningTotal,
                       setLoadingBuy,
                       setRunningTotal,
                       loadingProducts,
                       fetchProducts,
                       fetchCoins
                   }) => {
    const items = products || []; // If products is undefined, pass empty array
    return (
        <Table striped>
            <thead className="thead-dark">
            <tr>
                <th>Name</th>
                <th>Volume</th>
                <th>Quantity</th>
                <th>Price</th>
                <th style={{ textAlign: 'center' }}>Actions</th>
            </tr>
            </thead>
            <tbody>
            {!loadingProducts ? (
                !items || items.length <= 0 ? (
                    <tr>
                        <td colSpan="6" align="center">
                            <b>No Products yet</b>
                        </td>
                    </tr>
                ) : (
                    items.map(item => (
                        <tr key={item.id}>
                            <th scope="row">{item.name}</th>
                            <td>{item.volume}</td>
                            <td>{item.quantity}</td>
                            <th>R{item.price}</th>
                            <td align="center">
                                <div>
                                    <Button
                                        id={`${item.id}Button`}
                                        color="success"
                                        onClick={() =>
                                            checkout(item, runningTotal, setLoadingBuy, setRunningTotal, fetchProducts, fetchCoins)
                                        }
                                    >
                                        Buy
                                    </Button>
                                </div>
                            </td>
                        </tr>
                    ))
                )
            ) : (
                <div className="loader-container">
                    <img src="https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif" />
                </div>
            )}
            </tbody>
        </Table>
    );
};

export default DataTable;
