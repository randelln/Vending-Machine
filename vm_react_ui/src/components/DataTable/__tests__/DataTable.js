import React from 'react';
import { shallow } from 'enzyme';
import DataTableView, { checkout }  from "../DataTableView";
import Products from '../../../Pages/Home/__tests__/__data__/Products.json';

describe('DataTableView', () => {

    const props = {
        products: Products,
        runningTotal: 10,
        setLoadingBuy: jest.fn(),
        setRunningTotal: jest.fn(),
        loadingProducts: false,
        fetchProducts: jest.fn(),
        fetchCoins: jest.fn()
    }

    const view = shallow(<DataTableView {...props} />).dive();

    it('should render the DataTableView without issues', () => {
        expect(view).toMatchSnapshot();
    });

    it('should run checkout correctly', () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve({error: false, balance: 5 }),
            })
        );
        checkout(Products[2],10,props.setLoadingBuy, props.setRunningTotal, props.setRunningTotal, props.fetchProducts, props.fetchCoins)
        expect(props.setLoadingBuy).toHaveBeenCalledWith(true);
    });
});
