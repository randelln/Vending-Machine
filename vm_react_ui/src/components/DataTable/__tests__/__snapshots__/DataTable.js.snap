// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableView should render the DataTableView without issues 1`] = `
<table
  className="table table-striped"
>
  <thead
    className="thead-dark"
  >
    <tr>
      <th>
        Name
      </th>
      <th>
        Volume
      </th>
      <th>
        Quantity
      </th>
      <th>
        Price
      </th>
      <th
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="1"
    >
      <th
        scope="row"
      >
        Coke
      </th>
      <td />
      <td>
        10
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="1Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="2"
    >
      <th
        scope="row"
      >
        Fanta
      </th>
      <td />
      <td>
        2
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="2Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="3"
    >
      <th
        scope="row"
      >
        Sprite
      </th>
      <td />
      <td>
        3
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="3Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="4"
    >
      <th
        scope="row"
      >
        Lemon Twist
      </th>
      <td />
      <td>
        9
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="4Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="5"
    >
      <th
        scope="row"
      >
        Tab
      </th>
      <td />
      <td>
        5
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="5Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="6"
    >
      <th
        scope="row"
      >
        Pepsi
      </th>
      <td />
      <td>
        8
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="6Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="7"
    >
      <th
        scope="row"
      >
        KitKat
      </th>
      <td />
      <td>
        7
      </td>
      <th>
        R
        8
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="7Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="8"
    >
      <th
        scope="row"
      >
        Lays
      </th>
      <td />
      <td>
        12
      </td>
      <th>
        R
        5
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="8Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="9"
    >
      <th
        scope="row"
      >
        Simba
      </th>
      <td />
      <td>
        9
      </td>
      <th>
        R
        5
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="9Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
    <tr
      key="10"
    >
      <th
        scope="row"
      >
        Energade
      </th>
      <td />
      <td>
        11
      </td>
      <th>
        R
        10
      </th>
      <td
        align="center"
      >
        <div>
          <Button
            color="success"
            id="10Button"
            onClick={[Function]}
            tag="button"
          >
            Buy
          </Button>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
